### PR TITLE
Add Elastica\Query\Image

### PR DESCRIPTION
**Pull request by Jmoati**
Add support of https://github.com/kzwang/elasticsearch-image

### CODE REVIEW

**Comment by im-denisenko** [comment]
Is this plugin alive? I see they support ES 1.1.0, and last commit was half a year ago. 
PR with support for 1.4.2 was opened month ago and still has no comments. 
Elastica supports ES 1.4.3 right now, how would you use it with outdated plugin?

**Comment by Jmoati** [comment]
I use https://github.com/ifgh/elasticsearch-image which is a 1.4.2 fork.
I think that it will work with 1.4.2 too.

File: test/lib/Elastica/Test/Query/ImageTest.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    public function testToArrayFromImage()
+    {
+        $field = "image";
+
+        $query = new Image();
+        $query->setFieldFeature($field, 'CEDD');
+        $query->setFieldHash($field, 'BIT_SAMPLING');
+        $query->setFieldBoost($field, 100);
+
+        $query->setFieldImage($field, __DIR__ . '/../../../../data/test.jpg');
```
Comment (ruflin) [inline test/lib/Elastica/Test/Query/ImageTest.php]: To have a more readable path, check here: https://github.com/ruflin/Elastica/blob/master/test/lib/Elastica/Test/IndexTest.php#L124

File: lib/Elastica/Query/Image.php [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        return $this->setFieldParam($field, 'hash', $hash);
+    }
+
+    /**
+     * Set field image value
+     *
+     * @param  string $field
+     * @param  string $path File will be base64_encode
+     *
+     * @return Image
```
Comment (ruflin) [inline lib/Elastica/Query/Image.php]: I assume this should also be \Elastica\Query\Image

**Comment by ruflin** [comment]
To make sure it works with the current version, we should add an integration tests, which means we must add the plugin to our setup process. Like this we can also be sure it works with future versions.

Also we should add it then to the dependencies in the README file: https://github.com/ruflin/Elastica

@Jmoati If you need some help to get the setup done etc. we are here to help.

**Comment by im-denisenko** [comment]
More concrete steps how to add plugin to build:
- Add variable ES_IMAGE_PLUGIN_VER [here](https://github.com/ruflin/Elastica/blob/master/ansible/es-playbook.yml#L13)
- Add plugin installation [here](https://github.com/ruflin/Elastica/blob/master/ansible/roles/elasticsearch/tasks/main.yml#L67)
- Add plugin name [here](https://github.com/ruflin/Elastica/blob/master/ansible/roles/elasticsearch/templates/config-default.yml#L10)
- Add plugin to table of dependencies [here](https://github.com/ruflin/Elastica/blob/master/README.markdown)
- Implement tests
- Add **BIG BOLD** docblock that this query is not standard functionality of ES. Like [here](https://github.com/ruflin/Elastica/blob/master/lib/Elastica/Document.php#L159)

File: lib/Elastica/Query/Image.php (hunk 2 above)
Comment (im-denisenko) [inline lib/Elastica/Query/Image.php]: I think, since #774, here (and everywhere) should be `@return $this`

Comment (ruflin) [inline lib/Elastica/Query/Image.php]: :+1:

File: test/lib/Elastica/Test/Query/ImageTest.php (hunk 1 above)
Comment (Jmoati) [inline test/lib/Elastica/Test/Query/ImageTest.php]: Thx.

File: lib/Elastica/Query/Image.php (hunk 2 above)
Comment (Jmoati) [inline lib/Elastica/Query/Image.php]: I will replace by $this.

**Comment by Jmoati** [comment]
I will try this as soon as i can.

**Comment by Jmoati** [comment]
Can you please check.

File: ansible/roles/elasticsearch/tasks/main.yml [hunk 3]
```diff
@@ -45,6 +45,11 @@
     state=present
     update_cache=no
 
+- name: install image plugin
+  command: >
+    creates=/usr/share/elasticsearch/plugins/image
+    /usr/share/elasticsearch/bin/plugin --url https://github.com/SibaTokyo/elasticsearch-image/releases/download/{{ ES_IMAGE_PLUGIN_VER }}/elasticsearch-image-1.4.0.zip -install image
```
Comment (im-denisenko) [inline ansible/roles/elasticsearch/tasks/main.yml]: I guess "1.4.0" at the end of url should be variable too.

**Comment by im-denisenko** [comment]
All classes in `Elastica\Query` namespace can be instantiated by `Elastica\QueryBuilder`. You can add this query to querybuilder [here](https://github.com/ruflin/Elastica/blob/master/lib/Elastica/QueryBuilder/Version/Version140.php) and [here](https://github.com/ruflin/Elastica/blob/master/lib/Elastica/QueryBuilder/DSL/Query.php).

But this query isn't part of elasticsearch, so I'm not sure should it be added to querybuilder or not. 
Guess @ruflin or @webdevsHub knows better.

File: ansible/roles/elasticsearch/tasks/main.yml (hunk 3 above)
Comment (Jmoati) [inline ansible/roles/elasticsearch/tasks/main.yml]: You are right.

**Comment by Jmoati** [comment]
Maybe, the PR can be accepted before I implement it in QueryBuilder if necessary?

File: ansible/roles/elasticsearch/tasks/main.yml (hunk 3 above)
Comment (Jmoati) [inline ansible/roles/elasticsearch/tasks/main.yml]: It's fixed.

**Comment by ruflin** [comment]
Merged. Yes we can add it to the query builder in a second step.

**Comment by Jmoati** [comment]
Thx.

**Comment by webdevsHub** [comment]
I would prefer not to add it to Query DSL class because it is not an official elasticsearch query. Furthermore it will be difficult to update the Version classes with custom queries.

However we could implement a new "plugin" DSL in QueryBuilder and put all the custom plugin stuff in there

```
$qb = new QueryBuilder();
$qb->plugin()->image(/*...*/);

// or
$qb->elastica()->image(/*...*/);
```

**Comment by ruflin** [comment]
@webdevsHub I like the "plugin DSL" idea.

**Comment by webdevsHub** [comment]
I will look into that during weekends. I want to add some comments to the
QB unit tests anyway.

Are there any other plugin related queries/aggs already implemented in
Elastica that should be added to the new QB part?

**Comment by ruflin** [comment]
All the plugins are listed here in the README: https://github.com/ruflin/Elastica

Geocluster and attachment are the other two.

**Pull request merged**